src/Stream/BufferedStream.php

```diff
@@ -0,0 +1,264 @@
+<?php
+
+namespace Http\Message\Stream;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A buffered stream allow to buffer over an existing.
+ *
+ * You should use this decorator when you want to seek over a not seekable stream.
+ * This stream is however read-only.
```

Comment:
we could mention how this decorator buffers the stream and warn that really large streams could lead to memory problems. (or is it writing to disk if the stream is too large for memory?)